source/exe/signal_action.h

```diff
@@ -0,0 +1,118 @@
+#pragma once
+
+#include <signal.h>
+#include <unistd.h>
+
+#include "common/common/non_copyable.h"
+
+#include "server/backtrace.h"
+
+/**
+ * This class installs signal handlers for fatal signal types.
+ *
+ * These signals are handled:
+ *   SIGABRT
+ *   SIGBUS
+ *   SIGFPE
+ *   SIGILL
+ *   SIGSEGV
+ *
+ * Upon intercepting the signal the following actions are taken:
+ *
+ *   A Backtrace is printed from the address the signal was encountered at, if
+ *   it is possible to retrieve.
+ *
+ *   The signal handler is reset to the default handler (which is expected to
+ *   terminate the process).
+ *
+ *   The signal is raised again (which ultimately kills the process)
+ *
+ * The signal handler must run on an alternative stack so that we can do the
+ * stack unwind on the original stack. Memory is allocated for this purpose when
+ * this object is constructed. When this object goes out of scope the memory
+ * used for the alternate signal stack is destroyed and the default signal handler
+ * is restored.
+ *
+ * NOTE: Existing non-default signal handlers are overridden and will not be
+ * restored. If this behavior is ever required it can be implemented.
+ *
+ * It is recommended that this object be instantiated at the highest possible
+ * scope, eg, in main(). This enables fatal signal handling for almost all code
+ * executed.
+ */
+class SignalAction : NonCopyable {
+public:
+  SignalAction()
+      : guard_size_(sysconf(_SC_PAGE_SIZE)), altstack_size_(guard_size_ * 4), altstack_(nullptr) {
+    mapAndProtectStackMemory();
+    installSigHandlers();
+  }
+  SignalAction(const SignalAction&) = delete;
+  ~SignalAction() {
+    removeSigHandlers();
+    unmapStackMemory();
+  }
+  /**
+   * Helpers for testing guarded stack memory
+   */
+  void doGoodAccessForTest();
+  void tryEvilAccessForTest(bool end);
+
+private:
+  /**
+   * Allocate this many bytes on each side of the area used for alt stack.
+   *
+   * Set to system page size.
+   *
+   * The memory will be protected from read and write.
+   */
+  const size_t guard_size_;
+  /**
```

Comment:
Is there any reason NOT to use Doxygen formatted comments here? Doxygen can still produce useful documentation for private member vars depending on how it is configured to run.  I've found this information to be useful in past projects.